File: Lesson-2/anchor-tests/tests/instructions/initialize.ts
import { Environment } from "../environment";
import * as spl_token from '@solana/spl-token';
import { PublicKey } from '@solana/web3.js';
import * as anchor from '@coral-xyz/anchor';
import * as constants from "../constants";
import * as web3 from '@solana/web3.js';
import { assert } from 'chai';


interface Initialize {
    signer: anchor.web3.Keypair,
    vault: PublicKey,
    mint: anchor.web3.Keypair,
    metadata: PublicKey,
    open_time: number,
    input: number,
    name: string,
    symbol: string,
    uri: string,
}

export async function initialize(
    test_env: Environment,
    initialize: Initialize,
) {

    const mintParams = {
        name: initialize.name,
        symbol: initialize.symbol,
        uri: initialize.uri,

    }

    await test_env.program.methods.initialize(
        new anchor.BN(initialize.open_time),
        initialize.input,
        mintParams,
    ).accounts({
        signer: initialize.signer.publicKey,
        vault: initialize.vault,
        mint: initialize.mint.publicKey,
        mintMetadata: initialize.metadata,
        mplTokenMetadata: constants.MetaplexTokenMetadataProgram,
        tokenProgram: spl_token.TOKEN_PROGRAM_ID,
        associatedTokenProgram: spl_token.ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: web3.SystemProgram.programId
    }).signers([initialize.signer, initialize.mint]).rpc({ commitment: "confirmed" })

    let vaultData = await test_env.program.account.vault.fetch(initialize.vault);
    assert.strictEqual(vaultData.vaultContent, initialize.input);
    assert.strictEqual(vaultData.mint.toString(), initialize.mint.publicKey.toString());

}

File: Lesson-2/rust-tests/programs/rust-tests/tests/instructions/initialize.rs
use anchor_lang::prelude::{Pubkey, ToAccountMetas};

// Function to define the account metas needed for the Initialize instruction.
pub fn initialize_metas(
    signer: &Pubkey,
    mint: &Pubkey,
    vault: &Pubkey,
    metadata_address: &Pubkey,
) -> Vec<anchor_lang::solana_program::instruction::AccountMeta> {
    rust_tests::accounts::Initialize {
        signer: *signer,
        vault: *vault,
        mint: *mint,
        mint_metadata: *metadata_address,
        mpl_token_metadata: mpl_token_metadata::ID,
        token_program: anchor_spl::token::ID,
        associated_token_program: anchor_spl::associated_token::ID,
        system_program: solana_sdk::system_program::ID,
    }
    .to_account_metas(None)
}

// Function to define the data for the Initialize instruction.
pub fn initialize_data(
    open_time: i64,
    input: u8,
    name: &str,
    symbol: &str,
    uri: &str,
) -> rust_tests::instruction::Initialize {
    // Create a struct representing the parameters of the mint.
    let mint_parameters = rust_tests::state::MintParameters {
        name: name.to_string(),
        symbol: symbol.to_string(),
        uri: uri.to_string(),
    };

    // Return the Initialize instruction with the input and mint parameters.
    rust_tests::instruction::Initialize {
        open_time,
        input,
        mint_parameters,
    }
}
